src/alert-pcap.c

```diff
@@ -0,0 +1,406 @@
+        exit(EXIT_FAILURE);
+    }
+
+    output->pcap_writer = pcap_dump_open(output->pcap_file, filename);
+    if (output->pcap_writer == NULL) {
+        SCLogError(SC_ERR_LOGDIR_CONFIG, 
+                   "Cannot create alert pcap log output file %s: %s",
+                   filename, pcap_geterr(output->pcap_file));
+        exit(EXIT_FAILURE);
+    }    
+
+    output->output_file=pcap_dump_file(output->pcap_writer);
+    return output;
+}
+
+void AlertPcapFileClose(AlertPcapFile* output) {
+    
+    if (output->pcap_writer) {
+        pcap_dump_flush(output->pcap_writer);
+        pcap_dump_close(output->pcap_writer);
+    }
+    
+    if (output->pcap_file) {
+        pcap_close(output->pcap_file);
+    }
+}
+
+static TmEcode AlertPcapLogProcess(ThreadVars *t, Packet *p, void *thread_data, PacketQueue *pq,
+                 PacketQueue *postpq)
+{
+    char proto[16] = "", timebuf[64] = "";
+    char srcip[46] = "", dstip[46] = "", directory[PATH_MAX], filename[PATH_MAX];
+
+    AlertPcapLogThreadData *td = (AlertPcapLogThreadData *)thread_data;
+        
+    /* short circuit output if the packet doesn't contain an alert
+     * or the flow associated with the packet isn't supposed to have 
+     * an alert */
+    if (!(p->flow) || !(p->flow->flags & FLOW_TIMEMACHINE_ENABLED)) 
+        return TM_ECODE_OK;
+
+    CreateIsoTimeString(&p->flow->startts, timebuf, sizeof(timebuf));
+
+    if (FLOW_IS_IPV4(p->flow)) {
+        PrintInet(AF_INET, (const void*)&p->flow->src.addr_data32[0], srcip, sizeof(srcip));
+        PrintInet(AF_INET, (const void*)&p->flow->dst.addr_data32[0], dstip, sizeof(dstip));
+    } else if (FLOW_IS_IPV6(p->flow)) {
+        PrintInet(AF_INET6, (const void*)&p->flow->src.addr_data32, srcip, sizeof(srcip));
+        PrintInet(AF_INET6, (const void*)&p->flow->dst.addr_data32, dstip, sizeof(dstip));
+    }
+
+    if (SCProtoNameValid(p->flow->proto) == TRUE) {
+        strlcpy(proto, known_proto[p->flow->proto], sizeof(proto));
+    } else {
+        snprintf(proto, sizeof(proto), "%03" PRIu32, p->flow->proto);
+    }
+     
+    snprintf(directory, sizeof(directory), "%s/%.10s/%s-%s", 
+             td->apl_data->directory, timebuf, srcip, dstip);
+
+    if (p->flow->proto == IPPROTO_ICMP) {
+        snprintf(filename, sizeof(filename), "%s/%s-%s-%s.ICMP.cap", 
+                 directory, srcip, dstip, timebuf); 
+    } else {
+        snprintf(filename, sizeof(filename), "%s/%s:%hu-%s:%hu-%s.%s.cap", 
+                 directory, srcip, p->flow->sp, dstip, p->flow->dp, timebuf, proto);
+    }    
+       
+    /* see if we can find an output file */    
+    if (!(td->current_file) || strcmp(td->current_file->filename, filename) != 0) {
+        TAILQ_FOREACH(td->current_file, &td->pcap_files, next) {
+            if (strcmp(td->current_file->filename, filename) == 0) {
+                TAILQ_REMOVE(&td->pcap_files, td->current_file, next);
```

Comment:
If we find a file we can move it to the end since we know this file was recently written to.  That way when when we are attempting to figure out which files we no longer want to write to, we are guaranteeing the first file in the queue is the oldest so we shouldn't need to iterate through the entire queue every occurrence.